File: src/main/java/cz/wa/secretagent/game/action/ActionFactory.java
//package cz.wa.secretagent.game.action;
//
//import java.util.HashMap;
//import java.util.Map;
//
//import cz.wa.secretagent.game.sensor.SensorFactory;
//import cz.wa.secretagent.game.starter.MapStarter;
//import cz.wa.secretagent.world.entity.agent.AgentEntity;
//import secretAgent.game.ProjectileFactory;
//import secretAgent.game.action.AgentAction;
//import secretAgent.world.SamWorld;
//
///**
// * Creates actions with caching.
// *
// * @author Ondrej Milenovsky
// */
//public class ActionFactory {
//    private final AgentEntity agent;
//    private final MapStarter mapStarter;
//    private final ProjectileFactory projectileFactory;
//    private final SensorFactory sensorFactory;
//    private final SamWorld world;
//
//    private final Map<Object, AgentAction<AgentEntity>> cache;
//
//    public ActionFactory(AgentEntity agent, SamWorld world, SensorFactory sensorFactory,
//            MapStarter mapStarter, ProjectileFactory projectileFactory) {
//        this.agent = agent;
//        this.world = world;
//        this.sensorFactory = sensorFactory;
//        this.mapStarter = mapStarter;
//        this.projectileFactory = projectileFactory;
//        cache = new HashMap<Object, AgentAction<AgentEntity>>();
//    }
//
//    @SuppressWarnings("unchecked")
//    public <C> C getAction(Class<C> clazz) {
//        if (cache.containsKey(clazz)) {
//            return (C) cache.get(clazz);
//        } else {
//            C instance;
//            try {
//                instance = clazz.newInstance();
//            } catch (InstantiationException | IllegalAccessException e) {
//                throw new RuntimeException(e);
//            }
//            AgentAction<AgentEntity> object = (AgentAction<AgentEntity>) instance;
//            object.setAgent(agent);
//            object.setWorld(world);
//            object.setActionFactory(this);
//            object.setSensorFactory(sensorFactory);
//            object.init();
//            cache.put(clazz, object);
//            return instance;
//        }
//    }
//
//    public MapStarter getMapStarter() {
//        return mapStarter;
//    }
//
//    public ProjectileFactory getProjectileFactory() {
//        return projectileFactory;
//    }
//}

File: src/main/kotlin/secretAgent/game/action/ActionFactory.kt
package secretAgent.game.action

import cz.wa.secretagent.game.sensor.SensorFactory
import cz.wa.secretagent.game.starter.MapStarter
import cz.wa.secretagent.world.entity.agent.AgentEntity
import secretAgent.game.ProjectileFactory
import secretAgent.world.SamWorld
import java.util.HashMap

/**
 * Creates actions with caching.
 *
 * @author Ondrej Milenovsky
 */
/**
 * Creates actions with caching.
 *
 * @author Ondrej Milenovsky
 */
class ActionFactory(
        private val agent: AgentEntity,
        private val world: SamWorld,
        private val sensorFactory: SensorFactory,
        val mapStarter: MapStarter,
        val projectileFactory: ProjectileFactory?) {

    private val cache: MutableMap<Any, AgentAction<AgentEntity>> = HashMap()

    fun <C> getAction(clazz: Class<C>): C {
        if (cache.containsKey(clazz)) {
            return cache[clazz] as C
        } else {
            val instance: C
            try {
                instance = clazz.newInstance()
            } catch (e: InstantiationException) {
                throw RuntimeException(e)
            } catch (e: IllegalAccessException) {
                throw RuntimeException(e)
            }

            val `object` = instance as AgentAction<AgentEntity>
            `object`.agent = agent
            `object`.world = world
            `object`.actionFactory = this
            `object`.sensorFactory = sensorFactory
            `object`.init()
            cache[clazz] = `object`
            return instance
        }
    }
}


/**
 * Action to control some agent.
 * Most actions are on/off, if an agent calls fire(true), then he fires until called fire(false).
 *
 * @author Ondrej Milenovsky
 */
abstract class AgentAction<E : AgentEntity> {

    var agent: E? = null

    var world: SamWorld? = null
    //        internal set
    var actionFactory: ActionFactory? = null
    //        internal set
    var sensorFactory: SensorFactory? = null
//        internal set

    abstract fun init()
}